File: open_mahjong_unity/Assets/Scripts/GameScene/Calculation/CalculationScript/Classical/ClassicalCalc/ClassicalHepaiCheck.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace Classical {
    /// <summary>
    /// 古典麻将和牌检测与副番计算。
    /// 返回 (基础副, 总副, 副番名列表, 番名列表)。
    /// </summary>
    public class ClassicalHepaiCheck {
        private static readonly HashSet<int> Dragons = new HashSet<int> { 45, 46, 47 };
        private static readonly HashSet<int> Winds = new HashSet<int> { 41, 42, 43, 44 };
        private static readonly HashSet<int> Honors = new HashSet<int> { 41, 42, 43, 44, 45, 46, 47 };
        private static readonly HashSet<int> Terminals = new HashSet<int> { 11, 19, 21, 29, 31, 39 };
        private static readonly HashSet<int> Yaojiu;
        private static readonly HashSet<string> ManganFanSet = new HashSet<string> {
            "大三元", "大四喜", "小四喜", "天和", "地和", "九莲宝灯", "国士无双"
        };

        private static readonly Dictionary<string, int> FuValueMap = new Dictionary<string, int> {
            { "和牌", 10 }, { "自摸", 2 }, { "边嵌吊", 2 }, { "门前清", 2 },
            { "刻子", 2 }, { "暗刻", 4 }, { "明杠", 8 }, { "暗杠", 16 },
            { "幺九刻", 4 }, { "幺九暗刻", 8 }, { "幺九明杠", 16 }, { "幺九暗杠", 32 },
            { "番牌刻", 8 }, { "番牌暗刻", 16 }, { "番牌明杠", 32 }, { "番牌暗杠", 64 },
        };

        private readonly ClassicalCombinationSolver _solver = new ClassicalCombinationSolver();
        private readonly bool _debug;

        static ClassicalHepaiCheck() {
            Yaojiu = new HashSet<int>(Terminals);
            foreach (int h in Honors) Yaojiu.Add(h);
        }

        public ClassicalHepaiCheck(bool debug = false) {
            _debug = debug;
        }

        /// <summary>
        /// 和牌检测主入口，返回最优拆解的 (基础副, 总副, 副番名列表, 番名列表)。
        /// hand_list 应包含和牌张。
        /// </summary>
        public Tuple<int, int, List<string>, List<string>> HepaiCheck(
            List<int> handList, List<string> tilesCombination,
            List<string> wayToHepai, int getTile) {
            int completeStep = tilesCombination.Count * 3;
            var pt = new ClassicalPlayerTiles(handList, tilesCombination, completeStep);
            var doneList = new List<ClassicalPlayerTiles>();
            _solver.NormalCheck(pt, doneList);

            int bestBaseFu = 0, bestTotalFu = 0;
            List<string> bestFuFans = new List<string>();
            List<string> bestFans = new List<string>();

            foreach (var item in doneList) {
                _solver.BuildHandAndCombinationMapping(item);
                var (baseFu, totalFu, fuFanList, fanList) = FanCount(item, getTile, wayToHepai);
                if (totalFu > bestTotalFu) {
                    bestBaseFu = baseFu;
                    bestTotalFu = totalFu;
                    bestFuFans = fuFanList;
                    bestFans = fanList;
                }
            }
            return Tuple.Create(bestBaseFu, bestTotalFu, bestFuFans, bestFans);
        }

        /// <summary>
        /// 仅返回最大基础副数（不含番倍），供副数提示使用。
        /// </summary>
        public int FushuCheck(
            List<int> handList, List<string> tilesCombination,
            List<string> wayToHepai, int getTile) {
            int completeStep = tilesCombination.Count * 3;
            var pt = new ClassicalPlayerTiles(handList, tilesCombination, completeStep);
            var doneList = new List<ClassicalPlayerTiles>();
            _solver.NormalCheck(pt, doneList);

            int best = 0;
            foreach (var item in doneList) {
                _solver.BuildHandAndCombinationMapping(item);
                var (baseFu, _, _, _) = FanCount(item, getTile, wayToHepai);
                if (baseFu > best) best = baseFu;
            }
            return best;
        }

        private (int baseFu, int totalFu, List<string> fuFanTags, List<string> fanList) FanCount(
            ClassicalPlayerTiles pt, int getTile, List<string> wayToHepai) {
            NormalizeMeldVisibility(pt, getTile, wayToHepai);
            var combinationList = new List<string>(pt.CombinationList);
            var fanpaiSet = GetActiveFanpaiSet(wayToHepai);

            int fu = 0, fan = 0;
            var fuFanTags = new List<string>();
            var fanList = new List<string>();
            bool isMangan = false;

            if (wayToHepai.Contains("和牌")) { fu += FuValueMap["和牌"]; fuFanTags.Add("和牌"); }
            if (wayToHepai.Contains("自摸")) { fu += FuValueMap["自摸"]; fuFanTags.Add("自摸"); fanList.Add("自摸"); }
            if (wayToHepai.Contains("边嵌吊")) { fu += FuValueMap["边嵌吊"]; fuFanTags.Add("边嵌吊"); }

            var (triplets, pairTile) = ExtractSets(combinationList);
            var fuRepeatCount = new Dictionary<string, int>();
            foreach (var (sign, tile) in triplets) {
                var (addFu, tag) = CalcSetFu(sign, tile, fanpaiSet);
                fu += addFu;
                fuRepeatCount.TryGetValue(tag, out int cnt);
                fuRepeatCount[tag] = cnt + 1;
            }

            var handTiles = pt.HandTilesMapped.Count > 0
                ? new List<int>(pt.HandTilesMapped)
                : new List<int>();
            var suitSet = new HashSet<int>();
            bool hasHonor = false;
            foreach (int t in handTiles) {
                if (Honors.Contains(t)) hasHonor = true;
                else suitSet.Add(t / 10);
            }

            if (hasHonor && suitSet.Count == 1) { fan++; fanList.Add("混一色"); }
            else if (!hasHonor && suitSet.Count == 1) { fan += 3; fanList.Add("清一色"); }
            else if (handTiles.All(t => Honors.Contains(t))) { fan += 3; fanList.Add("字一色"); }

            if (triplets.Count > 0 && triplets.All(x => "KkGg".Contains(x.sign))) { fan++; fanList.Add("鸾凤和鸣"); }

            var dragonPungs = new HashSet<int>();
            var windPungs = new HashSet<int>();
            foreach (var (_, tile) in triplets) {
                if (Dragons.Contains(tile)) dragonPungs.Add(tile);
                if (Winds.Contains(tile)) windPungs.Add(tile);
            }

            if (dragonPungs.Count == 3) { fanList.Add("大三元"); isMangan = true; }
            else if (dragonPungs.Count == 2 && Dragons.Contains(pairTile)) { fan += 2; fanList.Add("小三元"); }

            if (windPungs.Count == 4) { fanList.Add("大四喜"); isMangan = true; }
            else if (windPungs.Count == 3 && Winds.Contains(pairTile)) { fanList.Add("小四喜"); isMangan = true; }

            if (wayToHepai.Contains("岭上开花") || wayToHepai.Contains("杠上开花")) { fan++; fanList.Add("岭上开花"); }
            if (wayToHepai.Contains("海底捞月")) { fan++; fanList.Add("海底捞月"); }
            if (wayToHepai.Contains("金鸡夺食") || wayToHepai.Contains("抢杠和")) { fan++; fanList.Add("金鸡夺食"); }
            if (wayToHepai.Contains("天和")) { fanList.Add("天和"); isMangan = true; }
            if (wayToHepai.Contains("地和")) { fanList.Add("地和"); isMangan = true; }
            if (wayToHepai.Contains("九莲宝灯")) { fanList.Add("九莲宝灯"); isMangan = true; }
            if (wayToHepai.Contains("国士无双")) { fanList.Add("国士无双"); isMangan = true; }

            int baseFu = Math.Min(fu, 300);
            int totalFu = Math.Min(baseFu * (1 << fan), 300);

            foreach (var kv in fuRepeatCount) {
                fuFanTags.Add(kv.Value == 1 ? kv.Key : $"{kv.Key}*{kv.Value}");
            }

            if (isMangan) {
                var manganOnly = fanList.Where(f => ManganFanSet.Contains(f)).ToList();
                return (baseFu, 300, fuFanTags, manganOnly);
            }
            return (baseFu, totalFu, fuFanTags, fanList);
        }

        private static (List<(char sign, int tile)> triplets, int pairTile) ExtractSets(List<string> combinationList) {
            var triplets = new List<(char, int)>();
            int pairTile = -1;
            foreach (string c in combinationList) {
                if (c.Length < 2) continue;
                char sign = c[0];
                if (!int.TryParse(c.Substring(1), out int tile)) continue;
                if ("kKgG".IndexOf(sign) >= 0) triplets.Add((sign, tile));
                else if (sign == 'q') pairTile = tile;
            }
            return (triplets, pairTile);
        }

        private static (int fu, string tag) CalcSetFu(char sign, int tile, HashSet<int> fanpaiSet) {
            bool isKong = sign == 'g' || sign == 'G';
            bool isConcealed = sign == 'K' || sign == 'G';
            bool isFanpai = fanpaiSet.Contains(tile);
            bool isYaojiu = !isFanpai && Yaojiu.Contains(tile);

            if (isKong) {
                if (isFanpai) { string n = isConcealed ? "番牌暗杠" : "番牌明杠"; return (FuValueMap[n], n); }
                if (isYaojiu) { string n = isConcealed ? "幺九暗杠" : "幺九明杠"; return (FuValueMap[n], n); }
                { string n = isConcealed ? "暗杠" : "明杠"; return (FuValueMap[n], n); }
            }
            if (isFanpai) { string n = isConcealed ? "番牌暗刻" : "番牌刻"; return (FuValueMap[n], n); }
            if (isYaojiu) { string n = isConcealed ? "幺九暗刻" : "幺九刻"; return (FuValueMap[n], n); }
            { string n = isConcealed ? "暗刻" : "刻子"; return (FuValueMap[n], n); }
        }

        private static HashSet<int> GetActiveFanpaiSet(List<string> wayToHepai) {
            var set = new HashSet<int>(Dragons);
            if (wayToHepai.Contains("门风东")) set.Add(41);
            else if (wayToHepai.Contains("门风南")) set.Add(42);
            else if (wayToHepai.Contains("门风西")) set.Add(43);
            else if (wayToHepai.Contains("门风北")) set.Add(44);
            return set;
        }

        private static void NormalizeMeldVisibility(ClassicalPlayerTiles pt, int getTile, List<string> wayToHepai) {
            bool isZimo = wayToHepai.Contains("妙手回春") || wayToHepai.Contains("自摸")
                || wayToHepai.Contains("杠上开花") || wayToHepai.Contains("岭上开花");
            if (isZimo) return;

            for (int i = 0; i < pt.CombinationList.Count; i++) {
                string comb = pt.CombinationList[i];
                if (comb == $"G{getTile}") {
                    if (pt.CombinationList.Contains($"S{getTile}") ||
                        pt.CombinationList.Contains($"S{getTile + 1}") ||
                        pt.CombinationList.Contains($"S{getTile - 1}"))
                        continue;
                    pt.CombinationList[i] = $"g{comb.Substring(1)}";
                    return;
                }
                if (comb == $"K{getTile}") {
                    if (pt.CombinationList.Contains($"S{getTile}") ||
                        pt.CombinationList.Contains($"S{getTile + 1}") ||
                        pt.CombinationList.Contains($"S{getTile - 1}"))
                        continue;
                    pt.CombinationList[i] = $"k{comb.Substring(1)}";
                    return;
                }
            }
        }
    }
}

File: open_mahjong_server/server/game_calculation/classical/classical_hepai_check.py
from typing import Dict, List, Tuple
from time import time
import logging

try:
    from .classical_solver import ClassicalCombinationSolver, PlayerTiles
except ImportError:
    from classical_solver import ClassicalCombinationSolver, PlayerTiles  # type: ignore

logger = logging.getLogger(__name__)


class Classical_Hepai_Check(ClassicalCombinationSolver):
    _DRAGONS = {45, 46, 47}  # 中白发
    _WINDS = {41, 42, 43, 44}  # 东南西北
    _HONORS = _WINDS | _DRAGONS  # 全部字牌
    _TERMINALS = {11, 19, 21, 29, 31, 39}  # 幺九数牌
    _YAOJIU = _TERMINALS | _HONORS  # 幺九牌集合（含字牌）
    _FANPAI = _DRAGONS  # 默认役牌：仅三元牌；门风通过传参动态加入
    _MANGAN_FAN_SET = {"大三元", "大四喜", "小四喜", "天和", "地和", "九莲宝灯", "国士无双"}  # 满贯番型
    # 副名 -> 单次副值映射（副番名仅保留名称，不拼接“x副”）。
    _FU_VALUE_MAP: Dict[str, int] = {
        "和牌": 10,
        "自摸": 2,
        "边嵌吊": 2,
        "门前清": 2,
        "刻子": 2,
        "暗刻": 4,
        "明杠": 8,
        "暗杠": 16,
        "幺九刻": 4,
        "幺九暗刻": 8,
        "幺九明杠": 16,
        "幺九暗杠": 32,
        "番牌刻": 8,
        "番牌暗刻": 16,
        "番牌明杠": 32,
        "番牌暗杠": 64,
    }

    def __init__(self, debug=False):
        self.debug = debug

    def debug_print(self, *args, **kwargs):
        if self.debug:
            logger.debug(*args, **kwargs)
            print(*args, **kwargs)

    def hepai_check(self, hand_list: list, tiles_combination, way_to_hepai, get_tile):
        """
        返回 (副数, 总副数, 副番名列表, 番名列表)。
        """
        complete_step = len(tiles_combination) * 3
        player_tiles = PlayerTiles(hand_list, tiles_combination, complete_step)
        check_done_list: List[PlayerTiles] = []
        self.normal_check(player_tiles, check_done_list)

        best_base_fu = 0
        best_total_fu = 0
        best_fu_fans: List[str] = []
        best_fans: List[str] = []
        if not check_done_list:
            return best_base_fu, best_total_fu, best_fu_fans, best_fans

        for item in check_done_list:
            # 枚举所有可行拆解并取总副数最高方案。
            mapped_hand, combination_str = self.build_hand_and_combination_mapping(item)
            item.hand_tiles_mapped = mapped_hand
            item.combination_str = combination_str
            base_fu, total_fu, fu_fan_list, fan_list = self.fan_count(item, get_tile, way_to_hepai)
            if total_fu > best_total_fu:
                best_base_fu = base_fu
                best_total_fu = total_fu
                best_fu_fans = fu_fan_list
                best_fans = fan_list

        return best_base_fu, best_total_fu, best_fu_fans, best_fans

    def hepai_check_all(self, hand_list: list, tiles_combination, way_to_hepai, get_tile):
        complete_step = len(tiles_combination) * 3
        player_tiles = PlayerTiles(hand_list, tiles_combination, complete_step)
        check_done_list: List[PlayerTiles] = []
        self.normal_check(player_tiles, check_done_list)

        result = []
        for item in check_done_list:
            mapped_hand, combination_str = self.build_hand_and_combination_mapping(item)
            item.hand_tiles_mapped = mapped_hand
            item.combination_str = combination_str
            result.append({
                "combination_str": combination_str,
                "hand_tiles_mapped": mapped_hand,
                "combination_list": item.combination_list[:],
            })
        return result

    def fan_count(self, player_tiles: PlayerTiles, get_tile, way_to_hepai):
        # 先按自摸/点和修正和牌张所在刻杠的明暗标记。
        self._normalize_meld_visibility(player_tiles, get_tile, way_to_hepai)
        combination_list = player_tiles.combination_list[:]
        fanpai_set = self._get_active_fanpai_set(way_to_hepai)  # 本局有效役牌（中白发+门风）
        fu = 0
        fan = 0
        fu_fan_tags: List[str] = []
        fan_list: List[str] = []
        is_mangan = False

        # 和牌副与边嵌吊副均按外部 way_to_hepai 传参计算。
        if "和牌" in way_to_hepai:
            fu += self._FU_VALUE_MAP["和牌"]
            fu_fan_tags.append("和牌")
        if "自摸" in way_to_hepai:
            fu += self._FU_VALUE_MAP["自摸"]
            fu_fan_tags.append("自摸")
            fan_list.append("自摸")
        if "边嵌吊" in way_to_hepai:
            fu += self._FU_VALUE_MAP["边嵌吊"]
            fu_fan_tags.append("边嵌吊")
        """ 
        暂时不计门前清
        if "门前清" in way_to_hepai:
            fu += self._FU_VALUE_MAP["门前清"]
            fu_fan_tags.append("门前清")
        """
        triplets, pair_tile = self._extract_sets(combination_list)
        fu_repeat_count: Dict[str, int] = {}
        for sign, tile in triplets:
            add_fu, tag = self._calc_set_fu(sign, tile, fanpai_set)
            fu += add_fu
            fu_repeat_count[tag] = fu_repeat_count.get(tag, 0) + 1

        hand_tiles = player_tiles.hand_tiles_mapped[:]
        if not hand_tiles:
            hand_tiles, _ = self.build_hand_and_combination_mapping(player_tiles)
        suits = {t // 10 for t in hand_tiles if t not in self._HONORS}
        has_honor = any(t in self._HONORS for t in hand_tiles)
        if has_honor and len(suits) == 1:
            fan += 1
            fan_list.append("混一色")
        elif (not has_honor) and len(suits) == 1:
            fan += 3
            fan_list.append("清一色")
        elif all(t in self._HONORS for t in hand_tiles):
            fan += 3
            fan_list.append("字一色")

        if all(s in {"K", "k", "G", "g"} for s, _ in triplets):
            fan += 1
            fan_list.append("鸾凤和鸣")

        dragon_pungs = {t for _, t in triplets if t in self._DRAGONS}
        wind_pungs = {t for _, t in triplets if t in self._WINDS}
        if len(dragon_pungs) == 3:
            fan_list.append("大三元")
            is_mangan = True
        elif len(dragon_pungs) == 2 and pair_tile in self._DRAGONS:
            fan += 2
            fan_list.append("小三元")

        if len(wind_pungs) == 4:
            fan_list.append("大四喜")
            is_mangan = True
        elif len(wind_pungs) == 3 and pair_tile in self._WINDS:
            fan_list.append("小四喜")
            is_mangan = True

        if "岭上开花" in way_to_hepai or "杠上开花" in way_to_hepai:
            fan += 1
            fan_list.append("岭上开花")
        if "海底捞月" in way_to_hepai:
            fan += 1
            fan_list.append("海底捞月")
        if "金鸡夺食" in way_to_hepai or "抢杠和" in way_to_hepai:
            fan += 1
            fan_list.append("金鸡夺食")
        if "天和" in way_to_hepai:
            fan_list.append("天和")
            is_mangan = True
        if "地和" in way_to_hepai:
            fan_list.append("地和")
            is_mangan = True
        if "九莲宝灯" in way_to_hepai:
            fan_list.append("九莲宝灯")
            is_mangan = True
        if "国士无双" in way_to_hepai:
            fan_list.append("国士无双")
            is_mangan = True

        # 结果总分封顶 300；命中满贯番时仅保留满贯番种输出。
        base_fu = min(fu, 300)
        total_fu = min(int(base_fu * (2 ** fan)), 300)

        for tag, cnt in fu_repeat_count.items():
            fu_fan_tags.append(tag if cnt == 1 else f"{tag}*{cnt}")

        if is_mangan:
            only_mangan_fans = [f for f in fan_list if f in self._MANGAN_FAN_SET]
            return base_fu, 300, fu_fan_tags, only_mangan_fans

        return base_fu, total_fu, fu_fan_tags, fan_list

    def fushucheck(self, hand_list: list, tiles_combination, way_to_hepai, get_tile) -> int:
        # 只返回可和牌拆解中的最大基础副数，不返回番型文本。
        best_base_fu = 0

        complete_step = len(tiles_combination) * 3
        player_tiles = PlayerTiles(hand_list, tiles_combination, complete_step)
        check_done_list: List[PlayerTiles] = []
        self.normal_check(player_tiles, check_done_list)
        for item in check_done_list:
            mapped_hand, combination_str = self.build_hand_and_combination_mapping(item)
            item.hand_tiles_mapped = mapped_hand
            item.combination_str = combination_str
            base_fu, _, _, _ = self.fan_count(item, get_tile, way_to_hepai)
            if base_fu > best_base_fu:
                best_base_fu = base_fu
        return best_base_fu

    def _extract_sets(self, combination_list: List[str]) -> Tuple[List[Tuple[str, int]], int]:
        # 从组合串中提取刻/杠列表与雀头牌值。
        triplets: List[Tuple[str, int]] = []
        pair_tile = -1
        for c in combination_list:
            if len(c) < 2:
                continue
            sign = c[0]
            try:
                tile = int(c[1:])
            except ValueError:
                continue
            if sign in {"k", "K", "g", "G"}:
                triplets.append((sign, tile))
            elif sign == "q":
                pair_tile = tile
        return triplets, pair_tile

    def _calc_set_fu(self, sign: str, tile: int, fanpai_set: set) -> Tuple[int, str]:
        # 按“普通/幺九/役牌”与“明暗/刻杠”计算单组副数和副名。
        is_kong = sign in {"g", "G"}
        is_concealed = sign in {"K", "G"}
        is_fanpai = tile in fanpai_set
        is_yaojiu = tile in self._YAOJIU and not is_fanpai

        if is_kong:
            if is_fanpai:
                name = "番牌暗杠" if is_concealed else "番牌明杠"
                return self._FU_VALUE_MAP[name], name
            if is_yaojiu:
                name = "幺九暗杠" if is_concealed else "幺九明杠"
                return self._FU_VALUE_MAP[name], name
            name = "暗杠" if is_concealed else "明杠"
            return self._FU_VALUE_MAP[name], name

        if is_fanpai:
            name = "番牌暗刻" if is_concealed else "番牌刻"
            return self._FU_VALUE_MAP[name], name
        if is_yaojiu:
            name = "幺九暗刻" if is_concealed else "幺九刻"
            return self._FU_VALUE_MAP[name], name
        name = "暗刻" if is_concealed else "刻子"
        return self._FU_VALUE_MAP[name], name

    def _get_active_fanpai_set(self, way_to_hepai: List[str]) -> set:
        # 根据 way_to_hepai 中门风参数动态生成役牌集合（不含圈风）。
        fanpai = set(self._FANPAI)
        if "门风东" in way_to_hepai:
            fanpai.add(41)
        elif "门风南" in way_to_hepai:
            fanpai.add(42)
        elif "门风西" in way_to_hepai:
            fanpai.add(43)
        elif "门风北" in way_to_hepai:
            fanpai.add(44)
        return fanpai

    def _normalize_meld_visibility(self, player_tiles: PlayerTiles, get_tile: int, way_to_hepai: List[str]) -> None:
        # 点和时将和牌张所在暗刻/暗杠降为明刻/明杠（暗转明）。
        zimo_or_not = any(i in ["妙手回春", "自摸", "杠上开花", "岭上开花"] for i in way_to_hepai)
        if not zimo_or_not:
            for comb in player_tiles.combination_list:
                if comb == f"G{get_tile}":
                    if any(i in player_tiles.combination_list for i in [f"S{get_tile}", f"S{get_tile+1}", f"S{get_tile-1}"]):
                        continue
                    player_tiles.combination_list.remove(comb)
                    player_tiles.combination_list.append(f"g{comb[1]}{comb[2]}")
                    return
                if comb == f"K{get_tile}":
                    if any(i in player_tiles.combination_list for i in [f"S{get_tile}", f"S{get_tile+1}", f"S{get_tile-1}"]):
                        continue
                    player_tiles.combination_list.remove(comb)
                    player_tiles.combination_list.append(f"k{comb[1]}{comb[2]}")
                    return


if __name__ == "__main__":
    test_save = [
    ["s37"], [
    11,11,11,
    21, 21, 21, 
    31, 31, 31, 
    39, 39
    ], 31, ["自摸"]]
    way_to_hepai = test_save[3]
    hepai_tile = test_save[2]
    tiles_list = test_save[1]
    combination_list = test_save[0]

    checker = Classical_Hepai_Check(debug=True)
    t0 = time()
    best = checker.hepai_check(tiles_list, combination_list, way_to_hepai, hepai_tile)
    all_patterns = checker.hepai_check_all(tiles_list, combination_list, way_to_hepai, hepai_tile)
    t1 = time()

    print("best_base_fu:", best[0])
    print("best_total_fu:", best[1])
    print("best_fu_fans:", best[2])
    print("best_fans:", best[3])
    print("all_patterns_count:", len(all_patterns))
    for idx, pattern in enumerate(all_patterns):
        print(idx, pattern)
    print("elapsed:", t1 - t0)

